File: interfaces/matlab/examples/ocp/dae_optimization_tutorial2/dae_optimization_tutorial2.m
clear;

BEGIN_ACADO;                                % Always start with "BEGIN_ACADO". 
    
    acadoSet('problemname', 'dae_optimization_tutorial2');% Set your problemname. If you 
                                                          % skip this, all files will
                                                          % be named "myAcadoProblem"
    
    
    DifferentialState       x;              % The differential state
    AlgebraicState          z;              % The algebraic state
    Control                 u;              % The control
    Parameter               p;              % The parameter
    
    
    %% Differential Equation
    f = acado.DifferentialEquation();       % Set the differential equation object
    
    f.add(dot(x) == -0.5*x-z+u*u     );     % Write down your equations. You can add 
    f.add(     0 ==  z+exp(z)+x-1.0+u);     % multiple ODE's and DAE's with f.add()   
                                            % when you have multiple states.
    

    
    %% Optimal Control Problem
    ocp = acado.OCP(0.0, 4.0);              % Set up the Optimal Control Problem (OCP)
                                            % Start at 0s, control in 10
                                            % intervals upto 5s
                                            
    ocp.minimizeMayerTerm(x*x + p*p);       % Minimize a Mayer term
    
    ocp.subjectTo( f );                     % Your OCP is always subject to your 
                                            % differential equation
                                            
    ocp.subjectTo( 'AT_START', x == 1.0 );  % Initial condition
    ocp.subjectTo( 'AT_END', x + p == 1.0 );% Terminal constraint  
    
    ocp.subjectTo( -1.0 <= x*u <= 1.0 );    % a path constraint
    
    
    
    %% Optimization Algorithm
    algo = acado.OptimizationAlgorithm(ocp); % Set up the optimization algorithm
    
    algo.set('KKT_TOLERANCE', 1e-5 );        % Set some parameters for the algorithm

    
    
END_ACADO;           % Always end with "END_ACADO".
                     % This will generate a file problemname_ACADO.m. 
                     % Run this file to get your results. You can
                     % run the file problemname_ACADO.m as many
                     % times as you want without having to compile again.



out = dae_optimization_tutorial2_RUN();     % Run the test. The name of the RUN file
                                            % is problemname_RUN, so in
                                            % this case
                                            % dae_optimization_tutorial2_RUN
                                            
draw;

File: examples/ocp/dae_optimization_tutorial2.cpp
/*
 *    This file is part of ACADO Toolkit.
 *
 *    ACADO Toolkit -- A Toolkit for Automatic Control and Dynamic Optimization.
 *    Copyright (C) 2008-2014 by Boris Houska, Hans Joachim Ferreau,
 *    Milan Vukov, Rien Quirynen, KU Leuven.
 *    Developed within the Optimization in Engineering Center (OPTEC)
 *    under supervision of Moritz Diehl. All rights reserved.
 *
 *    ACADO Toolkit is free software; you can redistribute it and/or
 *    modify it under the terms of the GNU Lesser General Public
 *    License as published by the Free Software Foundation; either
 *    version 3 of the License, or (at your option) any later version.
 *
 *    ACADO Toolkit is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *    Lesser General Public License for more details.
 *
 *    You should have received a copy of the GNU Lesser General Public
 *    License along with ACADO Toolkit; if not, write to the Free Software
 *    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */



 /**
  *    \file   examples/ocp/dae_optimization_tutorial2.cpp
  *    \author Boris Houska, Hans Joachim Ferreau
  *    \date   2010
  */


#include <acado_optimal_control.hpp>
#include <acado_gnuplot.hpp>


int main( ){

    USING_NAMESPACE_ACADO

    DifferentialState         x;             // definition of a differential state
    AlgebraicState            z;             // definition of an algebraic state
    Control                   u;             // definition of a control
    Parameter                 p;             // definition of a parameter
    DifferentialEquation      f;             // a differential equation


    f << dot(x) == -0.5*x-z+u*u;             // an example for a differential-
    f <<      0 ==  z+exp(z)+x-1.0+u;        // algebraic equation.

    OCP ocp( 0.0, 4.0 );                     // define an OCP with t_0 = 0.0 and T = 4.0
    ocp.minimizeMayerTerm( x*x + p*p );      // a Mayer term to be minimized

    ocp.subjectTo( f );                      // OCP should regard the DAE
    ocp.subjectTo( AT_START, x     == 1.0 ); // an initial value constraint
    ocp.subjectTo( AT_END  , x + p == 1.0 ); // an end (or terminal) constraint

    ocp.subjectTo( -1.0 <= x*u <= 1.0 );     // a path constraint

    OptimizationAlgorithm algorithm(ocp);    // define an algorithm
    algorithm.set( KKT_TOLERANCE, 1e-5 );    // define a termination criterion
	algorithm.solve();                       // to solve the OCP.

    return 0;
}



